File: vignettes/07_memory/csharp/07_memory.cs
// Memory and Context Providers — C#
//
// This sample demonstrates persistent memory using AIContextProvider:
//   1. A custom memory component that remembers user preferences.
//   2. Session-based state that persists across agent invocations.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using System.Text;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// Create the agent with a custom memory provider.
AIAgent agent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(new ChatClientAgentOptions
    {
        ChatOptions = new()
        {
            Instructions = "You are a friendly assistant. Keep answers brief. "
                + "Always address the user by name if you know it."
        },
        Name = "MemoryBot",
        AIContextProviders = [new UserPreferenceMemory()]
    });

// ── 1. Stateless behavior (new session each time) ──────────────────────
Console.WriteLine("=== Stateless (separate sessions) ===\n");
Console.WriteLine(await agent.RunAsync("My name is Alice."));
Console.WriteLine(await agent.RunAsync("What is my name?"));

// ── 2. Session-based memory (same session across calls) ────────────────
Console.WriteLine("\n=== Session-based Memory ===\n");
AgentSession session = await agent.CreateSessionAsync();

Console.WriteLine(await agent.RunAsync("My name is Alice.", session));
Console.WriteLine(await agent.RunAsync("What is my name?", session));

// ── 3. Custom DateTime context injection ───────────────────────────────
Console.WriteLine("\n=== DateTime Context Provider ===\n");

AIAgent datetimeAgent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(new ChatClientAgentOptions
    {
        ChatOptions = new()
        {
            Instructions = "You are a helpful assistant. Keep answers brief."
        },
        Name = "TimeAwareBot",
        AIContextProviders = [new DateTimeContextProvider()]
    });

AgentSession dtSession = await datetimeAgent.CreateSessionAsync();
Console.WriteLine(await datetimeAgent.RunAsync("What time is it right now?", dtSession));

// -------------------------------------------------------------------------- //
// Custom AIContextProvider: remembers user preferences in session state.      //
// -------------------------------------------------------------------------- //

/// <summary>
/// A simple memory component that extracts and remembers the user's name
/// from conversation messages, injecting it as context on subsequent calls.
/// </summary>
internal sealed class UserPreferenceMemory : AIContextProvider
{
    private readonly ProviderSessionState<UserPrefs> _state;

    public UserPreferenceMemory()
    {
        _state = new ProviderSessionState<UserPrefs>(
            _ => new UserPrefs(),
            nameof(UserPreferenceMemory));
    }

    public override IReadOnlyList<string> StateKeys => [_state.StateKey];

    protected override ValueTask<AIContext> ProvideAIContextAsync(
        InvokingContext context, CancellationToken ct = default)
    {
        var prefs = _state.GetOrInitializeState(context.Session);
        var sb = new StringBuilder();

        if (prefs.UserName is not null)
            sb.AppendLine($"The user's name is {prefs.UserName}.");
        else
            sb.AppendLine("You don't know the user's name yet.");

        return ValueTask.FromResult(new AIContext { Instructions = sb.ToString() });
    }

    protected override ValueTask StoreAIContextAsync(
        InvokedContext context, CancellationToken ct = default)
    {
        var prefs = _state.GetOrInitializeState(context.Session);

        // Simple extraction: look for "my name is X" in user messages.
        foreach (var msg in context.RequestMessages)
        {
            if (msg.Role != ChatRole.User) continue;
            var text = msg.Text ?? "";
            var idx = text.IndexOf("my name is", StringComparison.OrdinalIgnoreCase);
            if (idx >= 0)
            {
                var rest = text[(idx + 11)..].Trim().TrimEnd('.');
                var name = rest.Split(' ', StringSplitOptions.RemoveEmptyEntries)
                    .FirstOrDefault();
                if (name is not null)
                    prefs.UserName = name;
            }
        }

        _state.SaveState(context.Session, prefs);
        return ValueTask.CompletedTask;
    }
}

internal sealed class UserPrefs
{
    public string? UserName { get; set; }
}

// -------------------------------------------------------------------------- //
// DateTimeContextProvider: injects current date/time into system instructions //
// -------------------------------------------------------------------------- //

internal sealed class DateTimeContextProvider : AIContextProvider
{
    protected override ValueTask<AIContext> ProvideAIContextAsync(
        InvokingContext context, CancellationToken ct = default)
    {
        var now = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
        return ValueTask.FromResult(new AIContext
        {
            Instructions = $"Current date and time: {now}"
        });
    }
}

File: vignettes/07_memory/python/07_memory.py
"""
Memory and Context Providers — Python

This sample demonstrates persistent memory using context providers:
  1. InMemoryHistoryProvider for automatic conversation history.
  2. A custom context provider with before_run/after_run hooks.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama
"""

import asyncio
from datetime import datetime
from typing import Any

from agent_framework import BaseContextProvider, AgentSession, SessionContext
from agent_framework.ollama import OllamaChatClient


# --------------------------------------------------------------------------- #
# Custom context provider: injects the current date/time into instructions.   #
# --------------------------------------------------------------------------- #


class DateTimeProvider(BaseContextProvider):
    """Injects the current date and time as a system instruction."""

    DEFAULT_SOURCE_ID = "datetime"

    def __init__(self) -> None:
        super().__init__(self.DEFAULT_SOURCE_ID)

    async def before_run(
        self,
        *,
        agent: Any,
        session: AgentSession | None,
        context: SessionContext,
        state: dict[str, Any],
    ) -> None:
        now_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        context.extend_instructions(
            self.source_id,
            f"Current date and time: {now_str}",
        )

    async def after_run(
        self,
        *,
        agent: Any,
        session: AgentSession | None,
        context: SessionContext,
        state: dict[str, Any],
    ) -> None:
        pass  # Nothing to persist for this simple provider.


# --------------------------------------------------------------------------- #
# Main                                                                        #
# --------------------------------------------------------------------------- #


async def main() -> None:
    client = OllamaChatClient(
        host="http://localhost:11434",
        model_id="qwen3:8b",
    )

    # ── 1. Stateless agent (no memory) ────────────────────────────────────
    print("=== Stateless Agent ===")
    agent_no_memory = client.as_agent(
        name="ForgetfulBot",
        instructions="You are a helpful assistant. Keep answers brief.",
    )

    r1 = await agent_no_memory.run("My name is Alice.")
    print(f"Turn 1: {r1}")

    r2 = await agent_no_memory.run("What is my name?")
    print(f"Turn 2: {r2}")

    # ── 2. Agent with InMemoryHistoryProvider ─────────────────────────────
    print("\n=== Agent with Memory ===")
    from agent_framework import InMemoryHistoryProvider

    memory_agent = client.as_agent(
        name="MemoryBot",
        instructions="You are a helpful assistant. Keep answers brief.",
        context_providers=[InMemoryHistoryProvider()],
    )

    session = memory_agent.create_session()

    r1 = await memory_agent.run("My name is Alice.", session=session)
    print(f"Turn 1: {r1}")

    r2 = await memory_agent.run("What is my name?", session=session)
    print(f"Turn 2: {r2}")

    # ── 3. Agent with custom DateTimeProvider ─────────────────────────────
    print("\n=== Agent with Custom DateTimeProvider ===")
    datetime_agent = client.as_agent(
        name="TimeAwareBot",
        instructions="You are a helpful assistant. Keep answers brief.",
        context_providers=[DateTimeProvider(), InMemoryHistoryProvider()],
    )

    session = datetime_agent.create_session()
    r = await datetime_agent.run("What time is it right now?", session=session)
    print(f"Agent: {r}")


if __name__ == "__main__":
    asyncio.run(main())
